---
title: "Blood methylation pipeline using recountmethylation on EPIC array - V0_c"
author:
-  Liuhan Dai
date: "`r format(Sys.time(), '%d %B, %Y')`"
bibliography: bibliography.bib
csl: cell-numeric.csl
package: recountmethylation
output:
  BiocStyle::html_document:
    code_folding: show
    toc: yes
    tocfloat: yes
  BiocStyle::pdf_document: 
    toc: yes
    toc_depth: 2
vignette: > 
  %\VignetteIndexEntry{Blood_EPIC_v1}
  %\VignetteDepends{RCurl}
  %\usepackage[UTF-8]{inputenc} 
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

# Introdcution and overview: 

This script is to load, analyze and visualize `remethdb_epic-hm850k_h5_gm_0-0-3_1669220613.h5`. What it does is as follows:

1. load one compiled dataset stored on lab server.
  a. measured on illumina infinium methylation EPIC array
  b. Its url on online server is 
  https://recount.bio/data/remethdb_epic-hm850k_h5_gm_0-0-3_1669220613.h5

2. Subset blood and whole blood samples for each of them

3. Subset BCAT1 promoter cg probes for blood and whole blood subgroup

4. Visualization and statistical summary

# Environment setup
```{r setup, echo = FALSE, warning = FALSE}
suppressMessages(library(knitr))
suppressMessages(library(tidyverse))
suppressMessages(library(GenomicRanges))
suppressMessages(library(limma))
suppressMessages(library(minfi))
suppressMessages(library(ExperimentHub))
suppressMessages(library(recountmethylation))
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = TRUE,
                      message = TRUE)
```

# Load cached datasets on lab server

```{r}
BE.load <- function(fname,cachepath = "Z:/Liuhan Dai/DNA Methylation Detection/20230408 figure 6d") {
bfc <-BiocFileCache(cache = cachepath,ask = interactive())
if (grepl('C:',bfccache(bfc))) q()

filepath <- file.path(bfccache(bfc), fname)
# For this file, getdb() and get_rmdl() neither recoginize its file name.
# Therefore, decide to cache it manually
if (file.exists(filepath)) {
  if (!fname %in% bfcinfo(bfc)$rname) {
    bfcadd(bfc,fname,fpath =  filepath, action="asis")
  } else {
    print(paste(fname,'is already cached.'))
  }
} else {
  print(paste(fname,'doesn\'t exist.'))
}

# Load data by H5SE method
gm <- HDF5Array::loadHDF5SummarizedExperiment(file.path(bfccache(bfc), fname))
return(gm)
}

# load 'remethdb_epic-hm850k_h5_gm_0-0-3_1669220613.h5' as gm
gm <- BE.load('remethdb_epic-hm850k_h5_gm_0-0-3_1669220613.h5')
md <-  colData(gm) # get sample metadata (DO NOT CONVERT TO DATA FRAME!!!!! using as.data.frame)
```

# Subset blood and whole blood samples for `gm`

```{r}
# extract GSM ID subset for blood and whole blood in `gm`

md_blood <- md[grepl('blood',md$tissue),]
unique(md_blood$tissue)
md_whole_blood <- md[grepl('whole_blood',md$tissue),]
unique(md_whole_blood$tissue)
f <- function(x,y) nrow(md_whole_blood[y == x,])
sapply(unique(md_whole_blood$sex), f, y =  md_whole_blood$sex); sapply(unique(md_whole_blood$disease), f, y =  md_whole_blood$disease)
```


Then, let us subset GenomicMethylSet object for whole_blood subgroup using gsm
IDs we have in metadata.

```{r}
BE.subset <- function(gm, md_whole_blood) {
gm_whole_blood <- gm[, gm$gsm %in% md_whole_blood$gsm] # filt h5se object. Here gsm and other meta data is in "column"
# However, once you extract metadata by colData() -> md or md_whole_blood where its metadata now is stored in row
md_whole_blood <- md_whole_blood[order(match(md_whole_blood$gsm, gm_whole_blood$gsm)),] # Synchronize order of md_whole_blood and gr_whole_blood by gsm sequence

if (identical(gm_whole_blood$gsm, md_whole_blood$gsm)) {
    return(gm_whole_blood)
  } else {
    print('GSM seuqence order doesn\'t match')
  }
}

gm_whole_blood <- BE.subset(gm,md_whole_blood)
```

Secondly, let us extract DNAm Beta-value fractions for a subset of 3 probes that
are within BCAT1 promoter region.

# Subset BCAT1 promoter cg probes for blood and whole blood subgroup in `gm`

```{r}

BE.BCAT1 <- function(gm_whole_blood) {
# save bval table
bval <- getBeta(gm_whole_blood)
bval <- t(bval) # get transpose
cg_BCAT1 <- c('cg02765913', 'cg23036244', 'cg10764357')
f <- function (x,y)  y[grepl(x,y)]
cgf_BCAT1 <- sapply(cg_BCAT1, f, y = colnames(bval)) # search for which probe for BCAT1 promoter exists in this  dataset
cgf_BCAT1 <- as.character(cgf_BCAT1) ; cgf_BCAT1 <- cgf_BCAT1[nchar(cgf_BCAT1) == 10]
bval_BCAT1 <- bval[,colnames(bval) %in% cgf_BCAT1]  # extract a subset of 3 cg probes that are within BCAT1 promoter region
rownames(bval_BCAT1) <- gsub("\\..*", "", rownames(bval_BCAT1)) # format rownames
BCAT1 <- as.data.frame(bval_BCAT1);
if (identical(gm_whole_blood$gsm, rownames(BCAT1))) {
    BCAT1 <- bind_cols(BCAT1,as_tibble(colData(gm_whole_blood)))
  }
BCAT1 <- as_tibble(BCAT1)

return(BCAT1)
}

gm_BCAT1 <- BE.BCAT1(gm_whole_blood) 
```

To get a quantitative assessment, we can also create a summary table for gm_BCAT1

```{r}
cg_BCAT1 <- c('cg02765913', 'cg23036244', 'cg10764357')

BCAT1_Sum <- function(BCAT1, cg) {
  sex_summary <- BCAT1 %>%
  group_by(sex) %>%
  summarize(mean_value = mean(.data[[cg]]),
            median_value = median(.data[[cg]]),
            min = min(.data[[cg]]),
            max = max(.data[[cg]])
            )

disease_summary <-
  BCAT1 %>%
  group_by(disease) %>%
  summarize(mean_value = mean(.data[[cg]]),
            median_value = median(.data[[cg]]),
            min = min(.data[[cg]]),
            max = max(.data[[cg]])
  )
sex_summary
disease_summary

# diseaseSex_summary <-
#   BCAT1 %>%
#   group_by(disease, sex) %>%
#   summarize(mean_value = mean(.data[[cg]]),
#             median_value = median(.data[[cg]]),
#             min = min(.data[[cg]]),
#             max = max(.data[[cg]])
#   )
# 
# diseaseSex_summary
}

BCAT1_Sum(gm_BCAT1,cg_BCAT1[3])

```

Now, the question is how to come up with a reasonable explanation for my 30% BCAT1 in human blood. Here is what I am going to do. 

First, figure out probe binding sequence using snapgene alignment
Second, come up with mathematical relationship of methylation signals among three probes.
Thirdly, see whether they can be consistent with my measurement in any ways.